### DISCUSSION TITLE
Trigger /api/webhook with curl

### DISCUSSION THREAD

**Discussion by lukonjun** [argoproj/argo-cd#9032] [Q&A] [answered]
Hi there, I would like to trigger the /api/webhook endpoint of the application set controller with curl or any other cli tool.   
I following the documentation from here https://argocd-applicationset.readthedocs.io/en/stable/Generators-Git/ to build this curl command:
```
curl -X POST argocd-applicationset-controller:7000/api/webhook -H "Content-Type: application/json" -d '{"message":"lets trigger"}'
Unknown webhook event
```
When running curl however the logs of the application set controller show no extra syncing going on.

Any ideas?

**Comment by crenshaw-dev** [comment, marked as answer]
ApplicationSet ignores events that don't look like git events.

https://github.com/argoproj/argo-cd/blob/14cdfc77da479cfd8c95708311000ce53bf22b31/applicationset/utils/webhook.go#L110-L117

Try setting one of those headers to `push`.

**Reply by lukonjun** [reply]
Hi @crenshaw-dev, got one step further, I adapted the call
```
curl -kvX POST -H "Content-Type:application/json" -H "X-Github-Event:push" -d '{"mess
age":"hi"}' argocd-applicationset-controller:7000/api/webhook
``` 
Now I get an empty response from the server and can see an exception in the logs
```
2022/04/07 17:25:19 http: panic serving 192.168.37.36:45048: runtime error: slice bounds out of range [1:0]
goroutine 25302 [running]:
net/http.(*conn).serve.func1()
	/usr/local/go/src/net/http/server.go:1802 +0xb9
panic({0x2120240, 0xc0009ec078})
	/usr/local/go/src/runtime/panic.go:1047 +0x266
github.com/argoproj/applicationset/pkg/utils.getGitGeneratorInfo({0x2204d40, 0xc000583800})
	/workspace/pkg/utils/webhook.go:162 +0x509
github.com/argoproj/applicationset/pkg/utils.(*WebhookHandler).HandleEvent(0xc0008e1650, {0x2204d40, 0xc000583800})
	/workspace/pkg/utils/webhook.go:71 +0x4c
github.com/argoproj/applicationset/pkg/utils.(*WebhookHandler).Handler(0xc0008e1650, {0x25e2880, 0xc0001f00e0}, 0xc00017e800)
	/workspace/pkg/utils/webhook.go:129 +0x2b7
net/http.HandlerFunc.ServeHTTP(0x1, {0x25e2880, 0xc0001f00e0}, 0xc0001f00e0)
	/usr/local/go/src/net/http/server.go:2047 +0x2f
net/http.(*ServeMux).ServeHTTP(0x0, {0x25e2880, 0xc0001f00e0}, 0xc00017e800)
	/usr/local/go/src/net/http/server.go:2425 +0x149
net/http.serverHandler.ServeHTTP({0x25d25f0}, {0x25e2880, 0xc0001f00e0}, 0xc00017e800)
	/usr/local/go/src/net/http/server.go:2879 +0x43b
net/http.(*conn).serve(0xc0008be000, {0x25ef1d8, 0xc0003315f0})
	/usr/local/go/src/net/http/server.go:1930 +0xb08
created by net/http.(*Server).Serve
	/usr/local/go/src/net/http/server.go:3034 +0x4e8
```
However I dont really understand this  `runtime error: slice bounds out of range [1:0]`

**Reply by crenshaw-dev** [reply]
I imagine it's trying to parse the JSON as a github or gitlab webhook payload. I'd recommend testing with an example payload from your SCM of choice.

**Reply by rishabh625** [reply]
@lukonjun : As @crenshaw-dev mentioned u can try one of this json: https://github.com/argoproj/argo-cd/tree/master/applicationset/utils/testdata

**Reply by sspreitzer** [reply]
@rishabh625: Url has gone, mind posting the example here?
New url as of 18th Aug 2023: https://github.com/argoproj/argo-cd/blob/master/applicationset/webhook/testdata/github-commit-event.json

**Reply by messiahUA** [reply]
I'm on version 2.8.6 and sending this:
curl -X POST -H 'Content-type: application/json' -H 'X-GitHub-Event:push' -d '{"ref": "","repository": {"html_url":"https://user@dev.azure.com/org/project/_git/repo"}}' localhost:8080/api/webhook

There is a successful "Received push event repo" log entry and by looking at the code this seems to be all that is needed, but for some reason, nothing happens. As far as I see it should refresh the appset and I'm 100% sure that appset has the same URL. Am I missing something?

It's also strange that even if I manually refresh an appset it doesn't get the latest changes in the repo, but I suppose this is offtopic.

**Reply by dongdongheo** [reply]
Just go through this matter recently, my case is that I wanna trigger the Applicationset generation manually from Github Workflow instead of implementing Github Webhook.
I found out the below curl command along with the payload block working exactly as expected. Hope this be useful for anyone dealing with this same topic:

```
curl -X POST -H "Content-Type: application/json" \
-H "X-Github-Event:pull_request" \
-d '{
  "action": "actiontype",
  "repository": {
    "name": "myrepository",
    "owner": {
      "login": "myorg"
    },
    "url": "https://api.github.com/repos/myorg/myrepository"
  }
}' \
https://applicationset.example.com/api/webhook
```
